packages/coinstac-client-server/src/routes.js

```diff
@@ -0,0 +1,182 @@
+    return true;
+  } catch {
+    return h.response('Invalid token.').code(400).takeover();
+  }
+};
+
+module.exports = (server) => {
+  return [
+    {
+      method: 'POST',
+      path: '/pullImages',
+      options: {
+        pre: [
+          { method: validateToken },
+        ],
+        handler: async (req, h) => {
+          try {
+            const { payload: { images } } = req;
+
+            await pullImagesFromList(images);
+            return h.response('Pulled images.').code(200);
+          } catch (error) {
+            return h.response(error.message).code(400);
+          }
+        },
+      },
+    },
+    {
+      method: 'POST',
+      path: '/removeImages',
+      options: {
+        pre: [
+          { method: validateToken },
+        ],
+        handler: async (req, h) => {
+          try {
+            const { payload: { images } } = req;
+
+            await removeImagesFromList(images);
+            return h.response('Removed images.').code(200);
+          } catch (error) {
+            return h.response(error.message).code(400);
+          }
+        },
+      },
+    },
+    {
+      method: 'POST',
+      path: '/startPipeline/{clientId}',
+      options: {
+        pre: [
+          { method: validateToken },
+        ],
+        handler: async (req, h) => {
+          try {
+            const { payload: { run }, params: { clientId } } = req;
+            const { runId } = run;
+
+            let subscribed = false;
+            const dataQueue = [];
+
+            const subscriptionURL = `/pipelineResult/${runId}`;
+
+            server.subscription(subscriptionURL, {
+              onSubscribe: () => {
+                subscribed = true;
+
+                dataQueue.forEach((data) => {
+                  server.publish(subscriptionURL, data);
+                });
+              },
+              onUnsubscribe: () => {
+                subscribed = false;
+              },
+            });
+
+            const remotePipelineManager = await getDockerManager(clientId);
+
+            const computationImageList = run.pipelineSnapshot.steps
+              .map(step => step.computations
+                .map(comp => comp.computation.dockerImage))
+              .reduce((acc, val) => acc.concat(val), []);
+
+            await pullImagesFromList(computationImageList);
+
+            const managerResponse = remotePipelineManager.startPipeline({
+              clients: run.clients,
+              spec: run.pipelineSnapshot,
+              runId,
+              timeout: run.timeout,
+            });
+
+            const { pipeline, result, stateEmitter } = managerResponse;
+
+            function publishData(data) { // eslint-disable-line no-inner-declarations
+              if (subscribed) {
+                server.publish(subscriptionURL, data);
+              } else {
+                dataQueue.push(data);
```

Comment:
this might be better served being a stream, as it will back pressure automatically

Reply:
I cannot get you, @rssk